### PR TITLE
fix(loss): exclude mixed_type padding atoms from the training loss

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
## Problem

In the `mixed_type` data format, short frames are padded with `type = -1` ghost atoms up to a fixed `nloc`, and the real atom count varies per frame within a batch. The training loss normalized by the padded scalar `natoms` and took unmasked or cross-frame-pooled means, so ghost atoms diluted the force/atomic denominators and mis-normalized the extensive energy/virial/property terms. As a result a padded `[3-atom + 5-atom]` batch did not produce the same loss/gradient as processing the 3-atom and 5-atom frames separately. Only `mixed_type` batches are affected; non-mixed training was already exact.

## Fix

The per-atom mask (`atype >= 0`) now reaches the loss under the existing `model_dict["mask"]` convention: pt_expt already propagates it; the pt (torch.jit) backend recovers it from `atype` via a new `TaskLoss._inject_atom_mask` helper called from every pt loss `forward` (the exported forward drops the model's per-atom mask, so it is recovered training-side only — the exported artifact is untouched).

Every loss term is then normalized per frame so that a padded batch equals the grad-accumulation of the individual frames at their real sizes: per-atom terms (force, atom_ener, atom_pref, atomic dos/tensor, spin real-force, generalized force) use a per-frame masked mean; extensive terms (energy, virial, property) divide by the per-frame real atom count; global already-reduced terms (global dos/tensor) use a plain mean with the previous atom-count weighting dropped. Every change reduces exactly to the previous formula when the mask is all-ones, so non-mixed training is numerically identical (no-op to rounding). Covered across five shared loss types in both backends: `deepmd/dpmodel/loss/{ener,ener_spin,dos,tensor,property}.py` (which serve pt_expt) and `deepmd/pt/loss/{ener,ener_spin,dos,tensor,property}.py`.

Two additional fixes surfaced during the work: the extensive property normalization called `xp.sum(mask, -1)` with a positional axis, which raises `TypeError` under the array_api_compat torch namespace (every pt_expt extensive-property run) — now `axis=-1`; and `ener_spin`'s MAE energy and real-force terms were pre-existingly inconsistent with `ener.py` (they summed over frames without per-atom normalization) — they are now aligned with `ener.py`, which changes their non-mixed MAE loss values (a deliberate bug fix, see Known Limitations).

## Test

New `source/tests/common/dpmodel/test_loss_padding.py` and `source/tests/pt/test_loss_padding.py` assert, for every per-atom and extensive term of all five loss types in both backends, that a padded `[3+5]` batch loss equals the mean of the two frames processed separately, plus an all-ones-mask non-mixed no-op guard per term, and a torch-tensor path through the dpmodel property loss (which reproduces the positional-axis crash on the old form). An audit added invariant coverage for the generalized-force and spin magnetic-force terms and confirmed they are free of padding artifacts.

## Known limitations

The tf backend loss is unchanged and retains the same mixed_type behavior (follow-up). The pt-only losses `dens`, `population`, `denoise` are not covered (follow-up). `ener_spin`'s magnetic-force (`force_mag`) MAE term uses a sum over frames rather than a mean, so it does not satisfy the frame-average invariant — this is not a padding artifact (ghost atoms are correctly excluded via `mask_mag`), but a separate pre-existing MAE frame-normalization inconsistency, left for a follow-up decision. The `enable_atom_ener_coeff` path sums ghost atomic energies before the energy reduction (pre-existing; ghost atom_ener is ~0 by convention). Existing `mixed_type` trainings will not reproduce numerically — the new values are the correct ones. Ghost label forces are assumed ~0 by the dpdata convention; the mask makes the loss robust even if they are not.

## Summary by CodeRabbit

* **New Features**
  * Added automatic atom-mask injection so padded/multi-frame inputs with ghost atoms are handled consistently.

* **Bug Fixes**
  * Reworked masked loss normalization to use per-frame masked-mean reductions for energy, forces, virials, atom/property terms, DOS/CDF, tensor L2, and spin losses.
  * Standardized masked global DOS/CDF and global tensor L2 to use unweighted mean squared error.
  * Improved masking behavior for generalized-force projection and updated RMSE/MAE reporting accordingly.

* **Tests**
  * Added/extended gradient-accumulation and padding-mask invariance suites across dpmodel and pt backends, including atom-mask injection coverage.

### CODE REVIEW

**Review by njzjz — APPROVED** [review]
Approve but see above non-blocking comments.

**Comment by wanghan-iapcm** [comment]
Thanks for the review. Addressing the pre-merge notes:

- **Explicit tracking for the deferred follow-ups.** The unchanged TF backend loss is now tracked in #5760 and the pt-only \`dens\`/\`population\`/\`denoise\` losses in #5761, and both are referenced from the loss-padding test modules (commit b143f982c) so the scope boundary is no longer only in the PR body.
- **force_mag MAE debt** is tracked by the strict \`xfail\` tests added to both loss-padding suites (they self-heal to a failure once \`force_mag\` MAE switches to a frame-wise mean); the test docstrings now carry the scope note too.
- **User-visible ener_spin MAE change.** The PR now carries the \`breaking change\` label so it surfaces in the generated release notes, and it is documented under Known Limitations in the PR body.
- **CI.** The remaining Python/C++ shards are running; I will confirm them green before merge.

File: source/tests/common/dpmodel/test_loss_padding.py [hunk 1]
```diff
@@ -0,0 +1,2126 @@
...
+
+# ---------------------------------------------------------------------------
+# force_mag MSE grad-accum invariant (NM equal across frames, ghost atoms non-magnetic)
+#
+# NOTE: force_mag MAE (dpmodel) uses xp.sum over frames instead of xp.mean,
+# so MAE force_mag FAILS the invariant with a 2x factor when frames=2.
+# This is a pre-existing frame-normalization artifact independent of ghost-atom
+# masking. Ghost atoms are correctly excluded (mask_mag=0 there). The MAE
+# artifact is reported as NEEDS_CONTEXT in the audit report.
+# ---------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/common/dpmodel/test_loss_padding.py:2027]: Intentionally deferred here, not fixed. The `force_mag` MAE frame-normalization inconsistency is tracked by a strict `xfail(strict=True)` test in both loss-padding suites (this branch), which self-heals to a failure the moment `force_mag` MAE is switched to a frame-wise mean — a CI-enforced tracker that is harder to lose than a plain issue link. The test module docstrings also carry the scope note alongside the TF-backend (#5760) and dens/population/denoise (#5761) follow-ups.

File: deepmd/dpmodel/loss/dos.py [hunk 2]
```diff
@@ -175,15 +179,14 @@ def call(
...
             global_label = xp.reshape(label_dict["dos"], (-1, self.numb_dos))
             diff = global_pred - global_label
+            # idiom 3: global dos is already padding-invariant; plain mean suffices
+            l2_global_loss_dos = xp.mean(xp.square(diff))
             if "mask" in model_dict:
-                atom_num = xp.sum(model_dict["mask"], axis=-1, keepdims=True)
-                l2_global_loss_dos = xp.mean(
-                    xp.sum(xp.square(diff) * atom_num, axis=0) / xp.sum(atom_num)
+                atom_num = xp.mean(
+                    xp.astype(xp.sum(model_dict["mask"], axis=-1), diff.dtype)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/dpmodel/loss/dos.py:186]: Deferring this one. This is the display-only `rmse_global_*` metric, not the loss or gradient: the global loss term itself is now a plain per-frame-invariant mean (idiom 3), so training is unaffected, and the non-masked path is unchanged from master (`sqrt(mean(diff**2)) / natoms`). What you flag is only the *reported* masked RMSE using `mean(real_natoms)` instead of per-frame normalization — a reasonable batch summary rather than a correctness issue, and it only differs when frame sizes vary (i.e. it matches the non-masked metric exactly otherwise).

This PR is scoped to the masked-atom fix in the loss; changing the reported-metric definition (and mirroring it across CDF, `TensorLoss`, and both PT mirrors) is a self-contained display-metric refinement that fits better as a follow-up than folded in here. Happy to open a tracking issue for it if you want it recorded.

**Review by njzjz — APPROVED** [review]
